subprojects/base-services/src/main/java/org/gradle/internal/hash/HashCode.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.hash;
+
+import javax.annotation.Nullable;
+import java.io.Serializable;
+
+public class HashCode implements Serializable {
+    private static final int MIN_NUMBER_OF_BYTES = 4;
+    private static final int MAX_NUMBER_OF_BYTES = 255;
+    private static final char[] HEX_DIGITS = "0123456789abcdef".toCharArray();
+
+    private final byte[] bytes;
+
+    private HashCode(byte[] bytes) {
+        this.bytes = bytes;
+    }
+
+    static HashCode fromBytesNoCopy(byte[] bytes) {
+        return new HashCode(bytes);
+    }
+
+    public static HashCode fromBytes(byte[] bytes) {
+        // Make sure hash codes are serializable with a single byte length
+        if (bytes.length < MIN_NUMBER_OF_BYTES || bytes.length > MAX_NUMBER_OF_BYTES) {
+            throw new IllegalArgumentException(String.format("Invalid hash code length: %d bytes", bytes.length));
+        }
+        return fromBytesNoCopy(bytes.clone());
+    }
+
+    public static HashCode fromInt(int value) {
+        return fromBytesNoCopy(new byte[] {
+            (byte) (value >> 24),
```

Comment:
Why did you change the order of the bytes in one of your commits?

Reply:
It makes the resulting value easier to interpret (as the bytes are not reversed). Since we only use `fromInt()` in tests, this is a valuable trait.